subsys/bluetooth/controller/ll_sw/ull_conn_iso.c

```diff
@@ -181,15 +181,14 @@ void ull_conn_iso_cis_established(struct ll_conn_iso_stream *cis)
 
 void ull_conn_iso_done(struct node_rx_event_done *done)
 {
-	struct lll_conn_iso_group *lll = (void *)HDR_ULL2LLL(done->param);
-	struct ll_conn_iso_group *cig = (void *)HDR_LLL2EVT(lll);
+	struct lll_conn_iso_group *lll;
+	struct ll_conn_iso_group *cig;
 	uint32_t ticks_drift_minus;
 	uint32_t ticks_drift_plus;
 
-	/* Skip if CIG terminated by local host */
-	if (unlikely(lll->handle == 0xFFFF)) {
-		return;
-	}
```

Comment:
As `conn_cleanup` is now only called when processing done event, same should be followed for connected ISO too, that any cleanup should be initiated only on event done.

This means,

1. if there is a race where in `ticker_cb` has finished prepare and incremented reference count in the ULL context, then after which the done event is doing the cleanup, then this check is redundant as the handle is set to 0xFFFF when the done event is processed.
2. if there is a race where `done` event in the ULL context is being processed for initiating cleanup and `ticker_stop` operation has been enqueued. At this point lets assume `ticker_cb` executes due to ticker timeout happening just after ULL context processed the done event, then this is a case similar to (1) above wherein the event is placed into the pipeline and has the reference count incremented.
3. if the ticker stop....

Wait,  in case 2, pipeline has the event which is not flushed. Ok, the removed checks have to be back so that event does not proceed after the handle has been set to 0xFFFF.

Thank you for making me re-analyse the race. I will revert back these deleted code.

Reply:
Glad to help - would have been nice to get rid of them.